bindings/rust/bench/src/openssl.rs

```diff
@@ -0,0 +1,110 @@
+// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+// SPDX-License-Identifier: Apache-2.0
+
+use crate::{
+    harness::{CipherSuite, ConnectedBuffer, CryptoConfig, ECGroup, Mode, TlsBenchHarness},
+    CA_CERT_PATH, SERVER_CERT_CHAIN_PATH, SERVER_KEY_PATH,
+};
+use openssl::ssl::{
+    ErrorCode, Ssl, SslAcceptor, SslConnector, SslFiletype, SslMethod, SslStream, SslVersion,
+};
+use std::error::Error;
+
+pub struct OpenSslHarness {
+    client_conn: SslStream<ConnectedBuffer>,
+    server_conn: SslStream<ConnectedBuffer>,
+}
+
+impl OpenSslHarness {
+    /// Process handshake for one connection, treating blocking errors as `Ok`
+    fn handshake_conn(&mut self, mode: Mode) -> Result<(), Box<dyn Error>> {
+        match match mode {
+            Mode::Client => self.client_conn.connect(),
+            Mode::Server => self.server_conn.accept(),
+        } {
+            Ok(_) => Ok(()),
+            Err(err) => {
+                if err.code() != ErrorCode::WANT_READ {
+                    Err(err.into())
+                } else {
+                    Ok(())
+                }
+            }
+        }
+    }
+}
+
+impl TlsBenchHarness for OpenSslHarness {
+    fn new(crypto_config: &CryptoConfig) -> Result<Self, Box<dyn Error>> {
+        let client_buf = ConnectedBuffer::new();
+        let server_buf = client_buf.clone_inverse();
+
+        let cipher_suite = match crypto_config.cipher_suite {
+            CipherSuite::AES_128_GCM_SHA256 => "TLS_AES_128_GCM_SHA256",
+            CipherSuite::AES_256_GCM_SHA384 => "TLS_AES_256_GCM_SHA384",
+        };
+
+        let ec_key = match crypto_config.ec_group {
+            ECGroup::SECP256R1 => "P-256",
+            ECGroup::X25519 => "X25519",
+        };
+
+        let mut client_builder = SslConnector::builder(SslMethod::tls())?;
+        client_builder.set_ca_file(CA_CERT_PATH)?;
+        client_builder.set_min_proto_version(Some(SslVersion::TLS1_3))?;
+        client_builder.set_ciphersuites(cipher_suite)?;
+        client_builder.set_groups_list(ec_key)?;
+
+        let mut server_builder = SslAcceptor::mozilla_modern_v5(SslMethod::tls())?;
```

Comment:
What is `mozilla_modern_v5`?

Reply:
To make a server builder, OpenSSL requires a set of predefined safe defaults (Mozilla's standards), and I just chose fairly arbitrary which one to use. We override some of the defaults later anyways, so it doesn't matter too much for us. I added a comment to make it make more sense.

Reply:
Next question: Why do we override the defaults? If we override some of the defaults later anyways, why not just stick with the default?

Reply:
To get a builder from `SslAcceptor`, you have to choose a set of safe defaults, but it seems like making a builder from `SslContext` is much more bare bones with no defaults. However, using `Ssl`/`SslContext`/`SslContextBuilder` instead of `SslAcceptor`/`SslAcceptorBuilder` has this caveat on the docs:

>OpenSSL’s default configuration is insecure. It is highly recommended to use SslAcceptor rather than Ssl directly, as it manages that configuration.

The `SslAcceptorBuilder` is functionally very similar to a `SslContextBuilder` (in fact it implements `Deref<Target=SslContextBuilder>`), but my logic was that a typical user would use the `SslAcceptorBuilder` and so we should benchmark with that, in case there are defaults we don't override that might affect performance.

Reply:
Perhaps as an example, which defaults aren't safe? I just worry that opaquely adding extra checks might give an unfair penalty to OpenSSL. E.g. if we are adding extra checks or something.

Reply:
I looked at how it's implemented in OpenSSL and I think we override all of the defaults anyways to be even more strict? Since that's the case, we don't need the `SslAcceptorBuilder` and can just use a `SslContextBuilder`. Making that change now.